pkg/azurefile/controllerserver.go

```diff
@@ -311,25 +350,51 @@ func (d *Driver) getShareUrl(sourceVolumeID string) (azfile.ShareURL, error) {
 
 	accountKey, err := d.cloud.GetStorageAccesskey(accountName, resourceGroupName)
 	if err != nil {
-		return azfile.ShareURL{}, fmt.Errorf("could not find key for storage account(%s) under resource group(%s), err %v", accountName, resourceGroupName, err)
+		return azfile.ServiceURL{}, "", fmt.Errorf("could not find key for storage account(%s) under resource group(%s), err %v", accountName, resourceGroupName, err)
 	}
 
 	credential, err := azfile.NewSharedKeyCredential(accountName, accountKey)
 	if err != nil {
 		klog.Errorf("NewSharedKeyCredential(%s) in CreateSnapshot failed with error: %v", accountName, err)
-		return azfile.ShareURL{}, err
+		return azfile.ServiceURL{}, "", err
 	}
 
 	u, err := url.Parse(fmt.Sprintf(fileUrlTemplate, accountName, d.cloud.Environment.StorageEndpointSuffix))
 	if err != nil {
 		klog.Errorf("parse fileUrlTemplate error: %v", err)
-		return azfile.ShareURL{}, err
+		return azfile.ServiceURL{}, "", err
 	}
 	if u == nil {
-		return azfile.ShareURL{}, fmt.Errorf("url is nil")
+		return azfile.ServiceURL{}, "", fmt.Errorf("url is nil")
 	}
 
 	serviceURL := azfile.NewServiceURL(*u, azfile.NewPipeline(credential, azfile.PipelineOptions{}))
 
-	return serviceURL.NewShareURL(fileShareName), nil
+	return serviceURL, fileShareName, nil
+}
+
+func (d *Driver) snapshotExists(ctx context.Context, sourceVolumeID string, name string) (bool, azfile.ShareItem, error) {
+	serviceURL, fileShareName, err := d.getServiceURL(sourceVolumeID)
+	if err != nil {
+		return false, azfile.ShareItem{}, err
+	}
+
+	// List share snapshots.
+	listSnapshot, err := serviceURL.ListSharesSegment(ctx, azfile.Marker{}, azfile.ListSharesOptions{Detail: azfile.ListSharesDetail{Metadata: true, Snapshots: true}})
+	if err != nil {
+		return false, azfile.ShareItem{}, err
+	}
+	for _, share := range listSnapshot.ShareItems {
+		klog.V(2).Infof("Share(%s)", share.Name)
+		if share.Metadata["initiator"] == name {
+			if share.Name == fileShareName {
+				klog.V(2).Infof("Share(%s) snapshot(%s) Metadata(%v)", share.Name, *share.Snapshot, share.Metadata)
+				return true, share, nil
+			} else {
+				return true, azfile.ShareItem{}, fmt.Errorf("snapshot(%s) under file share(%s) already exists, but the SourceVolumeId is different", name, *share.Snapshot)
```

Comment:
`snapshot(%s) already exists, while current file share name(%s) does not equal to %s, SourceVolumeId(%s)`

Reply:
done.